dart/dynamics/InverseKinematics.h

```diff
@@ -0,0 +1,841 @@
+  static InverseKinematicsPtr create(JacobianNode* _node);
+
+  /// Copying is not allowed
+  InverseKinematics(const InverseKinematics&) = delete;
+
+  /// Assignment is not allowed
+  InverseKinematics& operator = (const InverseKinematics&) = delete;
+
+  /// Virtual destructor
+  virtual ~InverseKinematics();
+
+  /// Solve the IK Problem. By default, the Skeleton itself will retain the
+  /// solved joint configuration. If you pass in true for _resetConfiguration,
+  /// then the joint positions will be returned to their original configuration
+  /// after the problem is solved.
+  bool solve(bool _resetConfiguration = false);
+
+  /// Same as solve(bool), but the config vector will be filled with the solved
+  /// configuration.
+  bool solve(Eigen::VectorXd& config, bool _resetConfiguration = false);
+
+  /// Clone this IK module, but targeted at a new Node. Any Functions in the
+  /// Problem that inherit InverseKinematics::Function will be adapted to the
+  /// new IK module. Any generic optimizer::Function will just be copied over
+  /// by pointer instead of being cloned.
+  InverseKinematicsPtr clone(JacobianNode* _newNode) const;
+
+  /// This class should be inherited by optimizer::Function classes that have a
+  /// dependency on the InverseKinematics module that they belong to. If you
+  /// pass an InverseKinematics::Function into the Problem of an
+  /// InverseKinematics module, then it will be properly cloned whenever the
+  /// InverseKinematics module that it belongs to gets cloned. Any Function
+  /// classes in the Problem that do not inherit InverseKinematics::Function
+  /// will just be copied over by reference.
+  class Function
+  {
+  public:
+
+    /// Enable this function to be cloned to a new IK module.
+    virtual optimizer::FunctionPtr clone(InverseKinematics* _newIK) const = 0;
+
+    /// Virtual destructor
+    virtual ~Function() = default;
+  };
+
+  /// ErrorMethod is a base class for different ways of computing the error of
+  /// an InverseKinematics module.
+  class ErrorMethod : public common::Subject
+  {
+  public:
+
+    typedef std::pair<Eigen::Vector6d, Eigen::Vector6d> Bounds;
+
+    /// The Properties struct contains settings that are commonly used by
+    /// methods that compute error for inverse kinematics.
+    struct Properties
+    {
+      /// Default constructor
+      Properties(const Bounds& _bounds =
+            Bounds(Eigen::Vector6d::Constant(-DefaultIKTolerance),
+                   Eigen::Vector6d::Constant( DefaultIKTolerance)),
+
+          double _errorClamp = DefaultIKErrorClamp,
+
+          const Eigen::Vector6d& _errorWeights = Eigen::compose(
+            Eigen::Vector3d::Constant(DefaultIKAngularWeight),
+            Eigen::Vector3d::Constant(DefaultIKLinearWeight)));
+
+      /// Bounds that define the acceptable range of the Node's transform
+      /// relative to its target frame.
+      std::pair<Eigen::Vector6d, Eigen::Vector6d> mBounds;
```

Comment:
Actually, that's why I prefer `Allowance`; it sounds like a tolerance. From my understanding, the goal of inverse kinematics is to find a joint positions for a given **exact** target transformation (or other criteria) of the end-effector. That meas the error between the target transformation and the actual transformation should be ideally zero. But it is not possible in numerical way so we give some allowance.

However, if we see the region as a valid or feasible set of target transformation, then `Bounds` would work.